### ISSUE TITLE
Confusing diagnostic with colon followed by a macro

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#52981]
I ran across this today and it took  me a few seconds to figure out what was going on! This code:

```rust
fn main() {
    println!("x"): // note the accidental colon
    println!("y");
}
```

yields:

```
error: expected type, found `{`
 --> src/main.rs:3:5
  |
3 |     println!("y");
  |     ^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error
```

We can probably improve this error!

**Comment by JohnTitor** [comment]
Current output ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2198913065af948aa796d5130572ab2f)):

Stable
```
error: expected type, found `{`
 --> src/main.rs:3:5
  |
2 |     println!("x"): // note the accidental colon
  |                  - tried to parse a type due to this
3 |     println!("y");
  |     ^^^^^^^^^^^^^
  |     |
  |     expected type
  |     in this macro invocation
  |     this macro call doesn't expand to a type
  |
  = note: this warning originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Nightly
```
error: expected type, found `{`
 --> src/main.rs:3:5
  |
2 |     println!("x"): // note the accidental colon
  |                  - tried to parse a type due to this type ascription
3 |     println!("y");
  |     ^^^^^^^^^^^^^
  |     |
  |     expected type
  |     in this macro invocation
  |     this macro call doesn't expand to a type
  |
  = note: `#![feature(type_ascription)]` lets you annotate an expression with a type: `<expr>: <type>`
  = note: for more information, see https://github.com/rust-lang/rust/issues/23416
  = note: this warning originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

**Comment by Centril** [comment]
Current output seems pretty decent and explains the problem well.
